File: src/boot_animation.ts
#!/usr/bin/env node
/**
 * ToolNet CLI - Boot Animation & Interactive Prompt (TypeScript / Node / Bun)
 * Hiệu ứng: ASCII scan, neon corners, glitch text, gradient blocks, interactive prompt
 *
 * Usage:
 *   npx tsx boot_animation.ts
 *   bun boot_animation.ts
 */

import chalk from "chalk";
import readline from "node:readline";
import stringWidth from "string-width";

// === CONFIG & COLORS ===
export const PINK = chalk.hex("#e879f9");
export const BLUE = chalk.hex("#3b82f6");
export const CYAN = chalk.hex("#67e8f9");
export const PURPLE = chalk.hex("#a855f7");
export const MAGENTA = chalk.hex("#d946ef");
export const HIGHLIGHT = chalk.black.bgHex("#0ea5e9");
export const GREEN = chalk.hex("#22c55e");
export const GRAY = chalk.hex("#9ca3af");
export const DARK_BORDER = chalk.hex("#333333");

// TOOLNET ASCII 8x7 blocks per letter, 7 letters = 49 cols + spacing
export const ASCII_TOOLNET = [
  "  TTTTT   OOOOO   OOOOO   L       N   N   EEEEE   TTTTT  ",
  "    T    O     O O     O  L       NN  N   E         T    ",
  "    T    O     O O     O  L       N N N   EEEE      T    ",
  "    T    O     O O     O  L       N  NN   E         T    ",
  "    T    O     O O     O  L       N  NN   E         T    ",
  "    T     OOOOO   OOOOO   LLLLL   N   N   EEEEE     T    ",
];

export const GLITCH_CHARS = ["█", "▓", "▒", "░", "▚", "▞"];
export const GLITCH_BLOCKS = ["█", "▓", "▒", "▚"];

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/** Gradient across width: pink -> purple -> blue -> cyan */
export function colorForCol(colIdx: number, total: number): (s: string) => string {
  const ratio = colIdx / Math.max(total - 1, 1);
  if (ratio < 0.25) return PINK;
  if (ratio < 0.50) return PURPLE;
  if (ratio < 0.75) return BLUE;
  return CYAN;
}

/** Center text within a given visual width, accounting for ANSI codes */
export function center(text: string, width: number): string {
  const visibleLen = stringWidth(text);
  if (visibleLen >= width) return text;
  const leftPad = Math.floor((width - visibleLen) / 2);
  const rightPad = width - visibleLen - leftPad;
  return " ".repeat(leftPad) + text + " ".repeat(rightPad);
}

/** Add glitch offset characters to text */
export function glitchText(text: string, intensity = 3): string {
  let result = "";
  for (const ch of text) {
    if (ch === " ") {
      result += " ";
      continue;
    }
    if (Math.random() < 0.15 && intensity > 0) {
      const offsetCh = GLITCH_CHARS[Math.floor(Math.random() * GLITCH_CHARS.length)];
      const offsetColor = Math.random() < 0.5 ? BLUE : PURPLE;
      result += offsetColor(offsetCh) + PINK.bold(ch);
    } else {
      result += PINK.bold(ch);
    }
  }
  return result;
}

/** Render ASCII art with pixel-reveal progress (0.0 -> 1.0) */
export function renderAsciiFrame(progress: number): string[] {
  const totalCols = ASCII_TOOLNET[0].length;
  const revealCol = Math.floor(totalCols * progress);
  const rows: string[] = [];

  for (const row of ASCII_TOOLNET) {
    let line = "";
    for (let colIdx = 0; colIdx < row.length; colIdx++) {
      const ch = row[colIdx];
      if (colIdx > revealCol || ch === " ") {
        line += " ";
      } else {
        const block = Math.random() < 0.08
          ? GLITCH_BLOCKS[Math.floor(Math.random() * GLITCH_BLOCKS.length)]
          : "█";
        const color = colorForCol(colIdx, totalCols);
        line += color(block);
      }
    }
    rows.push(line);
  }
  return rows;
}

/** Show the folder trust prompt like GitHub Copilot CLI */
export async function showPrompt(options: { cwd?: string; interactive?: boolean } = {}): Promise<number> {
  const cwd = options.cwd ?? process.cwd();
  const isTty = Boolean(process.stdin.isTTY && process.stdout.isTTY && options.interactive !== false);

  console.log();
  const pathLabel = `  ${cwd}  `;
  const pWidth = stringWidth(pathLabel);
  console.log(DARK_BORDER("┌" + "─".repeat(pWidth) + "┐"));
  console.log(DARK_BORDER("│") + GRAY(pathLabel) + DARK_BORDER("│"));
  console.log(DARK_BORDER("└" + "─".repeat(pWidth) + "┘"));
  console.log();

  console.log(
    chalk.white(
      "ToolNet can read files in this folder and, with your permission,\n" +
      "edit them or run code and shell commands. It will remember\n" +
      "your permissions for the rest of this session.\n\n" +
      "Do you trust the files in this folder?"
    )
  );
  console.log();

  const choices = [
    "1. Yes",
    "2. Yes, and remember this folder for future sessions",
    "3. No (Esc)",
  ];

  if (!isTty) {
    for (let i = 0; i < choices.length; i++) {
      if (i === 0) {
        console.log(HIGHLIGHT(`> ${choices[i]}`));
      } else {
        console.log(`  ${choices[i]}`);
      }
    }
    console.log();
    console.log(chalk.dim("↑/↓ to navigate · enter to select · esc to cancel"));
    console.log();
    await sleep(200);
    console.log(GREEN.bold("✓ Trusted. ToolNet is ready."));
    return 0;
  }

  return new Promise<number>((resolve) => {
    let selected = 0;
    const linesCount = choices.length + 2;

    function renderChoices(first = false) {
      if (!first) {
        readline.moveCursor(process.stdout, 0, -linesCount);
      }
      for (let i = 0; i < choices.length; i++) {
        readline.cursorTo(process.stdout, 0);
        process.stdout.write("\x1B[2K");
        if (i === selected) {
          process.stdout.write(HIGHLIGHT(`> ${choices[i]}`) + "\n");
        } else {
          process.stdout.write(`  ${choices[i]}\n`);
        }
      }
      readline.cursorTo(process.stdout, 0);
      process.stdout.write("\x1B[2K\n");
      readline.cursorTo(process.stdout, 0);
      process.stdout.write("\x1B[2K" + chalk.dim("↑/↓ to navigate · enter to select · esc to cancel"));
    }

    renderChoices(true);

    const rl = readline.createInterface({ input: process.stdin, escapeCodeTimeout: 50 });
    readline.emitKeypressEvents(process.stdin, rl);
    if (process.stdin.isTTY) process.stdin.setRawMode(true);
    process.stdin.resume();

    function cleanup() {
      process.stdin.removeListener("keypress", onKeypress);
      if (process.stdin.isTTY) process.stdin.setRawMode(false);
      rl.close();
    }

    function onKeypress(_str: string, key: readline.Key) {
      if (key.name === "up" || key.name === "k") {
        selected = (selected - 1 + choices.length) % choices.length;
        renderChoices();
      } else if (key.name === "down" || key.name === "j") {
        selected = (selected + 1) % choices.length;
        renderChoices();
      } else if (key.name === "1") {
        selected = 0;
        renderChoices();
      } else if (key.name === "2") {
        selected = 1;
        renderChoices();
      } else if (key.name === "3") {
        selected = 2;
        renderChoices();
      } else if (key.name === "return" || key.name === "enter") {
        cleanup();
        console.log("\n");
        if (selected === 2) {
          console.log(chalk.red("Aborted."));
          process.exit(1);
        } else {
          console.log(GREEN.bold("✓ Trusted. ToolNet is ready."));
          resolve(selected);
        }
      } else if (key.name === "escape" || (key.ctrl && key.name === "c")) {
        cleanup();
        console.log("\n");
        console.log(chalk.red("Aborted."));
        process.exit(1);
      }
    }

    process.stdin.on("keypress", onKeypress);
  });
}

/** Main boot animation sequence */
export async function animateBoot(options: {
  interactive?: boolean;
  cwd?: string;
  version?: string;
} = {}): Promise<void> {
  const version = options.version ?? "1.0.0";
  const frameWidth = 59;
  const cornerTop = MAGENTA("┌" + "─".repeat(frameWidth) + "┐");
  const cornerBot = MAGENTA("└" + "─".repeat(frameWidth) + "┘");
  const versionLine = PURPLE.dim(`CLI Version ${version}`);

  process.stdout.write("\x1B[?25l"); // hide cursor

  // Phase 1: Build ASCII with scan (55 frames)
  let lastLinesCount = 0;
  for (let frame = 0; frame < 55; frame++) {
    const progress = frame / 54;
    const asciiRows = renderAsciiFrame(progress);
    const welcome = glitchText("Welcome to ToolNet", 3);

    const lines: string[] = [];
    lines.push(cornerTop);
    lines.push(center(welcome, frameWidth));
    lines.push("");

    const scanIdx = Math.floor(asciiRows.length * progress);
    for (let i = 0; i < asciiRows.length; i++) {
      if (i === scanIdx) {
        lines.push(MAGENTA("─".repeat(frameWidth)));
      } else {
        lines.push(center(asciiRows[i], frameWidth));
      }
    }

    lines.push("");
    lines.push(center(versionLine, frameWidth));
    lines.push(cornerBot);

    if (frame > 0) {
      readline.moveCursor(process.stdout, 0, -lastLinesCount);
    }
    for (const line of lines) {
      readline.cursorTo(process.stdout, 0);
      process.stdout.write("\x1B[2K" + line + "\n");
    }
    lastLinesCount = lines.length;

    await sleep(40 + Math.random() * 20); // slight jitter
  }

  // Phase 2: Hold with subtle glitch flicker (20 frames)
  for (let frame = 0; frame < 20; frame++) {
    const asciiRows = renderAsciiFrame(1.0);
    const welcome = glitchText("Welcome to ToolNet", 2);

    const lines: string[] = [];
    lines.push(cornerTop);
    lines.push(center(welcome, frameWidth));
    lines.push("");
    for (const row of asciiRows) {
      lines.push(center(row, frameWidth));
    }
    lines.push("");
    lines.push(center(versionLine, frameWidth));
    lines.push(cornerBot);

    readline.moveCursor(process.stdout, 0, -lastLinesCount);
    for (const line of lines) {
      readline.cursorTo(process.stdout, 0);
      process.stdout.write("\x1B[2K" + line + "\n");
    }
    lastLinesCount = lines.length;

    await sleep(80);
  }

  process.stdout.write("\x1B[?25h"); // show cursor

  // Phase 3: Prompt
  await showPrompt({ cwd: options.cwd, interactive: options.interactive });
}

// Auto-run when executed directly via bun or node/tsx
const isDirectRun =
  import.meta.url === `file://${process.argv[1]}` ||
  (typeof (import.meta as any).main === "boolean" && (import.meta as any).main) ||
  (process.argv[1] && process.argv[1].endsWith("boot_animation.ts"));

if (isDirectRun) {
  animateBoot().catch((err) => {
    process.stdout.write("\x1B[?25h");
    console.error("\nAborted.", err);
    process.exit(1);
  });
}

File: boot_animation.py
#!/usr/bin/env python3
"""
ToolNet CLI - Boot Animation & Interactive Prompt
Hiệu ứng: ASCII scan, neon corners, glitch text, gradient blocks
Requires: pip install rich
"""

import time
import random
import sys
from rich.console import Console, Group
from rich.text import Text
from rich.panel import Panel
from rich.align import Align
from rich.live import Live
from rich.style import Style

console = Console()

# === CONFIG ===
PINK = "#e879f9"
BLUE = "#3b82f6"
CYAN = "#67e8f9"
PURPLE = "#a855f7"
MAGENTA = "#d946ef"

# TOOLNET ASCII 8x7 blocks per letter, 7 letters = 49 cols + spacing
# Each letter: 7 cols wide, 8 rows tall
ASCII_TOOLNET = [
    "  TTTTT   OOOOO   OOOOO   L       N   N   EEEEE   TTTTT  ",
    "    T    O     O O     O  L       NN  N   E         T    ",
    "    T    O     O O     O  L       N N N   EEEE      T    ",
    "    T    O     O O     O  L       N  NN   E         T    ",
    "    T    O     O O     O  L       N  NN   E         T    ",
    "    T     OOOOO   OOOOO   LLLLL   N   N   EEEEE     T    ",
]

# Map each column to a color for gradient effect
COLORS = [PINK, PINK, PURPLE, BLUE, BLUE, CYAN, CYAN]

def color_for_col(col_idx: int, total: int) -> str:
    """Gradient across width: pink -> purple -> blue -> cyan"""
    ratio = col_idx / max(total - 1, 1)
    if ratio < 0.25:
        return PINK
    elif ratio < 0.5:
        return PURPLE
    elif ratio < 0.75:
        return BLUE
    else:
        return CYAN

def draw_corners():
    """Draw neon corner brackets"""
    corner_tl = Text.from_markup(f"[{MAGENTA}]┌{'─'*58}┐[/{MAGENTA}]")
    corner_bl = Text.from_markup(f"[{MAGENTA}]└{'─'*58}┘[/{MAGENTA}]")
    return corner_tl, corner_bl

def glitch_text(text: str, intensity: int = 3) -> Text:
    """Add glitch offset characters to text"""
    result = Text()
    for i, ch in enumerate(text):
        if ch == " ":
            result.append(" ")
            continue
        if random.random() < 0.15 and intensity > 0:
            offset_ch = random.choice(["█", "▓", "▒", "░", "▚", "▞"])
            result.append(offset_ch, style=Style(color=random.choice([BLUE, PURPLE]), blink=True))
            result.append(ch, style=Style(color=PINK, bold=True))
        else:
            result.append(ch, style=Style(color=PINK, bold=True))
    return result

def render_ascii_rows(progress: float) -> list[Text]:
    """Render ASCII art rows with pixel-reveal progress (0.0 -> 1.0)"""
    total_cols = len(ASCII_TOOLNET[0])
    reveal_col = int(total_cols * progress)
    rendered_rows = []

    for row in ASCII_TOOLNET:
        row_text = Text()
        for col_idx, ch in enumerate(row):
            if col_idx > reveal_col:
                row_text.append(" ")
            elif ch == " ":
                row_text.append(" ")
            else:
                if random.random() < 0.08:
                    block = random.choice(["█", "▓", "▒", "▚"])
                else:
                    block = "█"
                color = color_for_col(col_idx, total_cols)
                row_text.append(block, style=Style(color=color, bold=True))
        rendered_rows.append(row_text)
    return rendered_rows

def scanline() -> Text:
    """Horizontal scanline bar"""
    return Text.from_markup(f"[{MAGENTA}]{'─'*58}[/{MAGENTA}]")

def animate_boot():
    """Main boot animation sequence"""
    console.clear()

    corner_top, corner_bot = draw_corners()
    version = Text("CLI Version 1.0.0", style=Style(color=PURPLE, dim=True))

    with Live(console=console, refresh_per_second=30, screen=False) as live:
        # Phase 1: Build ASCII with scan
        for frame in range(0, 55):
            progress = frame / 54
            ascii_rows = render_ascii_rows(progress)
            welcome = glitch_text("Welcome to ToolNet")

            lines = []
            lines.append(corner_top)
            lines.append(Align.center(welcome, width=60))
            lines.append(Text(""))

            scan_idx = int(len(ascii_rows) * progress)
            for i, arow in enumerate(ascii_rows):
                if i == scan_idx:
                    lines.append(scanline())
                lines.append(Align.center(arow, width=60))

            lines.append(Text(""))
            lines.append(Align.center(version, width=60))
            lines.append(corner_bot)

            live.update(Group(*lines))
            time.sleep(0.04 + random.uniform(0, 0.02))

        # Phase 2: Hold with subtle glitch flicker
        for _ in range(20):
            ascii_rows = render_ascii_rows(1.0)
            lines = []
            lines.append(corner_top)
            lines.append(Align.center(glitch_text("Welcome to ToolNet", intensity=2), width=60))
            lines.append(Text(""))
            for arow in ascii_rows:
                lines.append(Align.center(arow, width=60))
            lines.append(Text(""))
            lines.append(Align.center(version, width=60))
            lines.append(corner_bot)
            live.update(Group(*lines))
            time.sleep(0.08)

    # Phase 3: Prompt
    show_prompt()

def show_prompt():
    """Show the trust prompt like GitHub Copilot CLI"""
    console.print()

    path_text = Text("/root", style=Style(color="#9ca3af"))
    path_panel = Panel(path_text, border_style="#333333", padding=(0, 2), width=12)
    console.print(path_panel)
    console.print()

    desc = (
        "ToolNet can read files in this folder and, with your permission,\n"
        "edit them or run code and shell commands. It will remember\n"
        "your permissions for the rest of this session.\n\n"
        "Do you trust the files in this folder?"
    )
    console.print(desc, style="white")
    console.print()

    options = [
        ("1. Yes", True),
        ("2. Yes, and remember this folder for future sessions", False),
        ("3. No (Esc)", False),
    ]

    for opt, active in options:
        if active:
            console.print(f"[black on #0ea5e9]> {opt}[/black on #0ea5e9]")
        else:
            console.print(f"  {opt}")

    console.print()
    console.print("↑/↓ to navigate · enter to select · esc to cancel", style="dim")
    console.print()

    with console.status("[bold cyan]Waiting for input...", spinner="dots"):
        time.sleep(1)
    console.print("[bold green]✓ Trusted. ToolNet is ready.[/bold green]")

if __name__ == "__main__":
    try:
        animate_boot()
    except KeyboardInterrupt:
        console.print("\n[red]Aborted.[/red]")
        sys.exit(1)
